Feature/avoid oom

NobodyXu opened the pull request:
`helpers::download` currently downloads the entire file and stores it in memory before writing it to disk.
This has the side-effect of if the file is super large, then it might run out of memory and panic.

## TODO
 - Implementing `download_and_extract` that downloads and extracts without storing the intermediate steps on disk.

Signed-off-by: Jiahao XU <Jiahao_XU@outlook.com>

passcod commented on src/helpers.rs:
```diff
@@ -210,3 +230,116 @@ pub trait Template: Serialize {
+    /// will be cancelled on failure.
+    handle: AutoAbortJoinHandle<io::Result<()>>,
+    tx: mpsc::Sender<Bytes>,
+}
```
Might as well put those two in their own files, split up helpers before it sprawls too much

NobodyXu replied:
I have the exact same thought as this!

So much content in `helpers.rs` that is better to be split across files.

NobodyXu replied:
Done!

passcod commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,32 @@
+name: Rust
```
Can you extract that in its own PR pls

NobodyXu replied:
Sure.

passcod commented on src/helpers/async_file_writer.rs:
```diff
@@ -0,0 +1,177 @@
+
+/// AsyncFileWriter removes the file if `done` isn't called.
+#[derive(Debug)]
+pub struct AsyncFileWriter(ScopeGuard<AsyncFileWriterInner, fn(AsyncFileWriterInner), Always>);
```
It would be useful to maybe rename this or add some more docs, because it's confusing as to what it does. There's unzipping? There's reading? There's what appears to be an async byte stream implemented over a channel?

NobodyXu replied:
Honestly, I'm not good at writing doc alone.

So perhaps you can keep asking me question about internals of `AsyncFileWriter` and I will answer it here while adding it to the doc.

passcod replied:
Well, I understand well enough how it works, I've read it, the issue here if treating it as opaque, looking at it from the outside, like me in 6 months, or anyone who wants to hack on this in the future, it's confusing.

Let me walk through a thought process, here:

- Ok, this is something called an async file writer. So, I'm expecting it writes a file, or maybe some files, asyncly. Ok.
- So, it takes a path, alright, that must be the path of the file it's writing,
- and a format, which looks like an archive format, so maybe it writes that file in this format,
- and a... desired outputs array? what?
- Uh, let's look at the documentation...
- it passes the `Bytes` you give it (wait but nowhere in the signature do you pass bytes, oh it's via the `write` method? uhh)
- to another thread via an mpsc (weird but okay, sure, why?)
- and decompresses and unpacks if needed.
- okay, so actually, it _reads_ from a possibly-compressed file, so, okay, the path must be to the file to read
- wait but you _give_ it bytes?
- so, it reads from a `Bytes` that you give it... and writes files to... a path? fine i guess
- and wait this cancellation bit... it "removes the file if done is not called"?
- which file? one of the ones it wrote? which one?
- is it a typo? is it all of the files it wrote?
- it can't be the possibly-compressed file, right, because I'm just giving it as bytes
- also, why am I giving it as bytes? shouldn't I just give it whole if I have it?
- ohhh, this is meant to have, like a stream of data coming in, right?
- weird, but okay

So, this is the outlook I'm looking to clarify here by docs and/or names.

(But hold on, there's a part 2!)

NobodyXu replied:
I renamed `AsyncFileWriter` to `AsyncExtracter`, it should be more straight forward of what it is doing.

NobodyXu replied:
I've added more doc, added part of the PR description to it and rename `AsyncExtracter::write` to `feed`.

passcod replied:
Part 2 is, okay, so what we have is a streaming I/O processor that consumes an archive (and also sometimes just a plain bin) and writes some specified files within it to disk. Except if the archive is zip, in which case it will write everything that's contained.

So, that's a bit complicated. Can we simplify that? Can we make:
- a streaming I/O processor
- that consumes an archive
- and writes the specified contained files to disk.

An ArchiveStreamExtractor, let's say.

And then we'll specify that you can't call it with a plain bin. In fact we could make our own `ArchiveFormat` enum that's split from `PkgFmt` because `PkgFmt` is about the binstall interface and metadata and URLs and templates, and this new `ArchiveFormat` is actually _just_ about how to unpack this archive stream we have here. `ArchiveFormat` will _not_ include `Bin`, that way the caller is forced to do something else for that, like directly write it to a file.

Also (unless I've misunderstood the caveats here), we'll uphold our promise of writing only specified files for `Zip` by postprocessing and deleting files that don't match. That way it's always clear and obvious what the output will be, even though it's a bit hacky on the inside at the moment (and who knows, maybe later we can improve and do partial extraction with `Zip`).

So, that's a good place to start.

And then there's part 3, which is some questions about the internals:
- Instead of the `write` and `done` methods, could we take an `impl Stream<Item = Result<Bytes>>` and iterate it ourselves then finalise (call done)?
- Instead of a new ReadableRx type, could we write the bytes into a [`DuplexStream`](https://docs.rs/tokio/latest/tokio/io/struct.DuplexStream.html)? That seems to have been designed for precisely this purpose.

passcod replied:
And then we don't really _need_ to have better docs, because our public API for this will look like:

```rust
pub async fn extract_archive_stream(
    stream: impl Stream<Item = Result<Bytes>>,
    output_dir: impl AsRef<Path>,
    file_filter: Option<Vec<Cow<Path>>>,
) -> Result<()>
```

NobodyXu replied:
> And then we'll specify that you can't call it with a plain bin.

I actually consider an advantage.

Without `AsyncExtracter`, you would do potentially blocking file writing on the async thread, which means blocking the tokio runtime (which is a bad practice).

> Also (unless I've misunderstood the caveats here), we'll uphold our promise of writing only specified files for Zip by postprocessing and deleting files that don't match. That way it's always clear and obvious what the output will be, even though it's a bit hacky on the inside at the moment (and who knows, maybe later we can improve and do partial extraction with Zip).

I don't think that part is necessary.
The `desired_outputs` is a best-effort: If you could filter the outputs, then great, it saves I/O and disk usage.
If not, then simply extract the whole content is ok.

> Instead of the write and done methods, could we take an impl Stream<Item = Result<Bytes>> and iterate it ourselves then finalise (call done)?

Yeah, definitely could do.

> Instead of a new ReadableRx type, could we write the bytes into a [DuplexStream](https://docs.rs/tokio/latest/tokio/io/struct.DuplexStream.html)? That seems to have been designed for precisely this purpose.

No.

Currently, we read directly from the `Bytes` received and it requires one copy.
Adding `DuplexStream` adds another buffer and another copy.

NobodyXu replied:
> And then we don't really _need_ to have better docs, because our public API for this will look like:
> 
> ```rust
> pub async fn extract_archive_stream(
>     stream: impl Stream<Item = Result<Bytes>>,
>     format: ArchiveFormat,
>     output_dir: impl AsRef<Path>,
>     file_filter: Option<Vec<Cow<Path>>>,
> ) -> Result<()>
> ```

Done!

NobodyXu commented:
@passcod Regarding the pipelining to `zip` part, I planned on opening another PR so that the `ZipArchiver` can read bytes out of the file while the bytes are being appended into the file.

It should be doable as `ZipArchiver` only reads and the other parts only appends to the file.

passcod commented:
> Without AsyncExtracter, you would do potentially blocking file writing on the async thread, which means blocking the tokio runtime (which is a bad practice).

I mean, fair enough, but let's at least consider having two functions in the public API for this module, one for writing a file directly from a `Stream<Bytes>`, and the other for doing the extracting logic, that way it's clear what's going on.

> I don't think that part is necessary. The desired_outputs is a best-effort

Alright, sounds okay in that interpretation.

NobodyXu commented:
> I mean, fair enough, but let's at least consider having two functions in the public API for this module (that call the same underlying things, to be clear), one for writing a file directly from a Stream<Bytes>, and the other for doing the extracting logic, that way it's clear what's going on.

Yep, that's indeed much better that way.

NobodyXu commented:
@passcod Let's merge it for now and then open new PR to improve it incrementally.

I have a few other PRs that I want to do and I want to interleave between them.
Been doing this PR for 2 days I want to do some other things before going back to it.

A summary of what could be/need to be improved:
 - Adjust the size of the `mpsc::channel` to keep the extracter busy while avoiding reading in too much (OOM) and gives feedback to the server. Fix could be dynamically decide the buffer based on `fmt: PkgFmt` since some extracters (`PkgFmg::Bin` and `PkgFmt::Zip`) do less work on the critical path and need bigger buffer size.
 - The aforementioned of improvement "the ZipArchiver can read bytes out of the file while the bytes are being appended into the file."
 - Split API `extract_archive_stream`: Create a separate API for `PkgFmt::Bin`.

The following aspect, however, does not need improvement as of now:

> and who knows, maybe later we can improve and do partial extraction with Zip.

I added the `desired_outputs` for `find_crate_cratesio` since we download the whole crate but actually only needs the `Cargo.toml`.

Unless crates.io has plan to switch to using zip or there are other use cases for it, I suggest that having `desired_outputs` support for all tars is enough.
